Fully deprecate variadic inputs of checkpoint_sequential

sublee opened the pull request:
To support variadic inputs of `checkpoint_sequential` was deprecated at #21006. This case should be warned with `DeprecationWarning` for PyTorch 1.2, but it should be simply failed with `TypeError` since PyTorch 1.3. This patch removes the `DeprecationWarning` for PyTorch 1.2.

sublee committed:
Deprecate variadic arguments of checkpoint_sequential

sublee committed:
Sync docstring of checkpoint_sequential

sublee committed:
Test TypeError instead of DeprecationWarning

soumith commented:
@pytorchbot rebase this please

sublee commented:
@soumith Thank you for reviewing it.

yf225 commented:
@sublee We can't land it yet because test is failing:
```
Sep 24 20:52:17 ======================================================================
Sep 24 20:52:17 ERROR: test_checkpoint_module_list_multiple_args (__main__.TestCheckpoint)
Sep 24 20:52:17 ----------------------------------------------------------------------
Sep 24 20:52:17 Traceback (most recent call last):
Sep 24 20:52:17   File "test_utils.py", line 188, in test_checkpoint_module_list_multiple_args
Sep 24 20:52:17     torch.randn(1, 60, requires_grad=True)
Sep 24 20:52:17   File "test_utils.py", line 69, in _check_checkpoint_sequential
Sep 24 20:52:17     out = checkpoint_sequential(model_to_compare, num_chunks, *detached_inputs)
Sep 24 20:52:17   File "/opt/conda/lib/python3.6/site-packages/torch/utils/checkpoint.py", line 217, in checkpoint_sequential
Sep 24 20:52:17     preserve_rng_state=preserve_rng_state)
Sep 24 20:52:17   File "/opt/conda/lib/python3.6/site-packages/torch/utils/checkpoint.py", line 155, in checkpoint
Sep 24 20:52:17     return CheckpointFunction.apply(function, preserve, *args)
Sep 24 20:52:17   File "/opt/conda/lib/python3.6/site-packages/torch/utils/checkpoint.py", line 62, in forward
Sep 24 20:52:17     if preserve_rng_state:
Sep 24 20:52:17 RuntimeError: bool value of Tensor with more than one value is ambiguous
```

soumith reviewed (changes requested):
please fix test

sublee committed:
Make preserver_rng_state keyword-only

sublee committed:
checkpoint_sequential doesn't check if input is tuple

sublee committed:
Fix checkpoint test failures

sublee commented:
@yf225, @soumith Sorry for my mistake. I ran only the test cases I added at #21006 but I had not to do. I just fixed those failures. Please review it.

sublee commented:
@soumith I guess you lost the last update. Would you please review this PR again?

sublee committed:
Bump target version to v1.4

It was intended to deploy with v1.3.0 but late.

sublee commented:
This patch was intended to deploy with v1.3.0. However, today PyTorch v1.3.0 has been released. I update the version to v1.4 in the deprecation documentation and rebase this branch onto the current `master`.

albanD commented on test/test_utils.py:
```diff
@@ -44,41 +44,37 @@ def _check_checkpoint_sequential(
             for name, param in model.named_parameters()
         }
-        input_grad_not_checkpointed = [i.grad.data.clone() for i in inputs]
+        input_grad_not_checkpointed = input.grad.data.clone()
```
Can you use `.detach()` here instead of `.data` please?

sublee replied:
Sure! I agree with you.

albanD commented on test/test_utils.py:
```diff
@@ -44,41 +44,37 @@ def _check_checkpoint_sequential(
                 for name, param in model.named_parameters()
             }
-            input_grad_checkpointed = [d.grad.data.clone() for d in detached_inputs]
+            input_grad_checkpointed = detached.grad.data.clone()
```
Same data -> detach()

sublee replied:
I applied your suggestion: https://github.com/pytorch/pytorch/pull/25985/commits/01bf4641093a9396aac90976ee4bd4c215d412cc

albanD reviewed (approved):
Details in the test to make it look slightly better.
Then I'll merge.

Thanks for the PR !

sublee committed:
data -> detach() in test_utils

@albanD requested:

https://github.com/pytorch/pytorch/pull/25985#discussion_r353389420
https://github.com/pytorch/pytorch/pull/25985#discussion_r353389517

albanD commented:
Just rebased on top of master for CI.

zou3519 commented:
Hmm, If I am understanding this correctly it is not BC breaking because it enables behavior that used to throw a TypeError
